packages/arb-token-bridge-ui/src/components/App/App.tsx

```diff
@@ -172,21 +176,74 @@ const Injector = ({ children }: { children: React.ReactNode }): JSX.Element => {
       const l1Address = await l1Signer.getAddress()
       const l2Address = await l2Signer.getAddress()
 
-      try {
-        const l1AddressIsEOA = await addressIsEOA(l1Address, l1Provider)
-        const l2AddressIsEOA = await addressIsEOA(l2Address, l2Provider)
+      const walletType: WalletType = await (async () => {
+        const l1AddressIsSmartContract = await addressIsSmartContract(
+          l1Address,
+          l1Provider
+        )
+        const l2AddressIsSmartContract = await addressIsSmartContract(
+          l2Address,
+          l2Provider
+        )
+
+        if (!l1AddressIsSmartContract && !l2AddressIsSmartContract) {
+          return WalletType.EOA
+        }
+
+        if (l1AddressIsSmartContract && l2AddressIsSmartContract) {
+          // check if gnosis / gnosis like wallet w/ same L1 & L2 address & same owners
+          if (l1Address !== l2Address) {
+            console.warn(
+              `SC wallet error: wallets have different addresses: l1: ${l1Address}; l2: ${l2Address} `
+            )
+            return WalletType.UNSUPPORTED_CONTRACT_WALLET
+          }
 
-        if (!l1AddressIsEOA || !l2AddressIsEOA) {
-          actions.app.setConnectionState(ConnectionState.NOT_EOA)
-          return undefined
+          try {
+            const l1Contract = new Contract(
+              l1Address,
+              gnosisInterface,
+              l1Provider
+            )
+            const l1Owners: string[] = [
+              ...(await l1Contract.getOwners())
+            ].sort()
+
+            const l2Contract = new Contract(
+              l2Address,
+              gnosisInterface,
+              l2Provider
+            )
+            const l2Owners: string[] = [
+              ...(await l2Contract.getOwners())
+            ].sort()
+
+            const walletsHaveSameOwners =
+              JSON.stringify(l1Owners) === JSON.stringify(l2Owners)
+
+            if (!walletsHaveSameOwners) {
+              console.warn(
+                `SC wallet error; owners don't match:`,
+                l1Owners,
+                l2Owners
+              )
+            }
+
+            return walletsHaveSameOwners
+              ? WalletType.SUPPORTED_CONTRACT_WALLET
+              : WalletType.UNSUPPORTED_CONTRACT_WALLET
```

Comment:
do we need to `setConnectionState` if we are returning `UNSUPPORTED_CONTRACT_WALLET`